source/common/grpc/google_async_client_impl.cc

```diff
@@ -267,41 +344,28 @@ void GoogleAsyncStreamImpl::metadataTranslate(
   }
 }
 
-void GoogleAsyncStreamImpl::completionThread() {
-  ENVOY_LOG(debug, "completionThread running");
-  void* tag;
-  bool ok;
-  while (cq_.Next(&tag, &ok)) {
-    ENVOY_LOG(trace, "completionThread CQ event {} {}", operation(tag), ok);
-    parent_.dispatcher_.post([ this, op = operation(tag), ok ] { handleOpCompletion(op, ok); });
-  }
-  ENVOY_LOG(debug, "completionThread exiting");
+void GoogleAsyncStreamImpl::deferredDelete() {
+  ENVOY_LOG(debug, "Deferred delete");
+  tls_.unregisterStream(this);
+  dispatcher_.deferredDelete(std::unique_ptr<GoogleAsyncStreamImpl>(this));
```

Comment:
`std::unique_ptr<FooType>(this)` looks really wrong. Are we confident nothing else can hold a pointer or reference to this object? A comment here would make it look less shocking.

Reply:
There seem to be cases in the Envoy code base where there is a `unique_ptr<GoogleAsyncStreamImpl>`. Am I missing something or couldn't this cause a double free?

Reply:
Added comment; we only get here after cleanup(), which hands the object self-ownership of memory. After this method completes, nothing should be invoked on the stream, it's dead.